### PR TITLE
DNM: Add scanning page

### LINKED ISSUE #318: Dintinguish between candidates and sources
Add a flag to sources, to identify some sources as candidates.

These can be removed after X amount of time in the database if not flagged as a source.

### PR DESCRIPTION
**Pull request by acrellin**
This PR adds a candidates table, `/api/candidates` API endpoint handlers, and a scanning page with date, group and unsaved only/all candidates filters. When saving a candidate as a source, the user is presented with a list of groups to select from that will be associated with the new source. Spectra and photometry added to either candidates or sources are also added to the associated source/candidate wherever relevant. A daily cron job is added that removes old candidates not flagged as sources.

Closes #318

### CODE REVIEW

**Comment by acrellin** [comment]
[ch679]

**Comment by acrellin** [comment]
[ch680]

**Comment by acrellin** [comment]
[ch682]

**Comment by acrellin** [comment]
[ch749]

**Comment by stefanv** [comment]
@acrellin This looks good, but it's pretty hard to review: an enormous PR!  Is there any way to split out some of these changes?

In principle, the only problem I see is that we cannot rely on pg_cron, since we won't have that extension available on AWS or Cloud SQL.  I suggest that, instead, we write an extra service in baselayer, monitored by supervisord, that allows us to do scheduled jobs.

I've implemented a [proof of concept cron service](https://github.com/cesium-ml/baselayer/pull/86) and [an example of how to use it in SkyPortal](https://github.com/skyportal/skyportal/pull/376).

**Comment by acrellin** [comment]
@stefanv I've removed the pg_cron section and have added a daily job to delete old candidates. Any other changes you'd like to see?

**Comment by stefanv** [comment]
@acrellin yes, like I said before: this is a lot to review in one go. Any way we can add components one by one, and split out infrastructure changes? Currently, the PR runs many tens of pages, so it is hard to know which changes go toward what, and to separate concerns.

**Comment by acrellin** [comment]
@stefanv these changes all go together and are related to adding candidates to the data model -- this required updating the handlers for all objects associated with sources. The front-end additions implement the scanning page features.... What difference does it make to review a bunch of different PRs (each of which can't really stand on its own without breaking the existing functionality) vs one big one? It's the same amount of code to review, and I think it's actually easier to have it all centralized. It's a big PR because it's a big new feature that touches many parts of the code base.

**Comment by acrellin** [comment]
@stefanv if you have a specific proposal around what to pull out into a separate PR, I'm happy to take a look at that.

**Comment by stefanv** [comment]
I'm not trying to be petty. But I either need guidance on how to review 
this, or smaller PRs. If you can provide a more detailed overview of what's 
happening here, which parts you are more and less certain about, etc., then 
that could help. As is, it's very overwhelming to just see many, many pages 
of diffs.

File: config.yaml.defaults [hunk 1]
```diff
@@ -31,3 +31,11 @@ server:
 
 services:
     dask: False
+
+misc:
```
Comment (acrellin) [inline config.yaml.defaults]: The changes to this file are just for the cron job

File: jobs/delete_unsaved_candidates.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env python
```
Comment (acrellin) [inline jobs/delete_unsaved_candidates.py]: The changes to this file comprise the new cron job

File: skyportal/handlers/api/candidate.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
+import datetime
```
Comment (acrellin) [inline skyportal/handlers/api/candidate.py]: This is the handler file for the new `/api/candidates` endpoint

File: skyportal/handlers/api/comment.py [hunk 4]
```diff
@@ -31,8 +31,11 @@ def get(self, comment_id):
         comment = Comment.query.get(comment_id)
         if comment is None:
             return self.error('Invalid comment ID.')
-        # Ensure user/token has access to parent source
-        s = Source.get_if_owned_by(comment.source.id, self.current_user)
+        # Ensure user/token has access to parent source/candidate
```
Comment (acrellin) [inline skyportal/handlers/api/comment.py]: The changes to this file are due to the fact that the parent may now either be a comment or a candidate, so we add checks here

File: skyportal/handlers/api/internal/plot.py [hunk 5]
```diff
@@ -6,23 +6,37 @@
 # TODO this should distinguish between "no data to plot" and "plot failed"
 class PlotPhotometryHandler(BaseHandler):
     @auth_or_token
-    def get(self, source_id):
-        docs_json, render_items, custom_model_js = plot.photometry_plot(source_id)
+    def get(self, source_or_candidate_id):
```
Comment (acrellin) [inline skyportal/handlers/api/internal/plot.py]: The changes to this file account for the fact that plot data can be associated with either a source and a candidate, and some syntax changes (`black`)

File: skyportal/handlers/api/internal/source_views.py [hunk 6]
```diff
@@ -3,51 +3,61 @@
...
-    DBSession, Source, SourceView, GroupSource
-)
+from ....models import DBSession, Source, SourceView, GroupSource
 
 
-default_prefs = {
-    'maxNumSources': 10,
-    'sinceDaysAgo': 7
-}
+default_prefs = {"maxNumSources": 10, "sinceDaysAgo": 7}
```
Comment (acrellin) [inline skyportal/handlers/api/internal/source_views.py]: The changes to this file are just syntax changes (`black`)

File: skyportal/handlers/api/photometry.py [hunk 7]
```diff
@@ -1,22 +1,14 @@
-import os
-import io
-import base64
-from pathlib import Path
-import tornado.web
+import arrow
```
Comment (acrellin) [inline skyportal/handlers/api/photometry.py]: The non-syntax changes to this file are due to the fact that the parent may now either be a comment or a candidate, so we add checks here

File: skyportal/handlers/api/source.py [hunk 8]
```diff
@@ -1,20 +1,25 @@
-import tornado.web
 from sqlalchemy.orm import joinedload
-from sqlalchemy import func, desc
+from sqlalchemy import func
```
Comment (acrellin) [inline skyportal/handlers/api/source.py]: Syntax changes (`black`)

File: skyportal/handlers/api/spectrum.py [hunk 9]
```diff
@@ -1,9 +1,7 @@
-import tornado.web
-from sqlalchemy.orm import joinedload
 from marshmallow.exceptions import ValidationError
 from baselayer.app.access import permissions, auth_or_token
 from ..base import BaseHandler
-from ...models import DBSession, Spectrum, Comment, Instrument, Source
+from ...models import DBSession, Spectrum, Instrument, Source, Candidate
```
Comment (acrellin) [inline skyportal/handlers/api/spectrum.py]: The changes to this file are due to the fact that the parent may now either be a comment or a candidate, so we add checks here

File: skyportal/handlers/api/thumbnail.py [hunk 10]
```diff
@@ -6,7 +6,7 @@
 from PIL import Image
 from baselayer.app.access import permissions, auth_or_token
 from ..base import BaseHandler
-from ...models import DBSession, Photometry, Source, Thumbnail
+from ...models import DBSession, Candidate, Photometry, Source, Thumbnail
```
Comment (acrellin) [inline skyportal/handlers/api/thumbnail.py]: The changes to this file are due to the fact that the parent may now either be a comment or a candidate, so we add checks here

File: skyportal/models.py [hunk 11]
```diff
@@ -1,17 +1,22 @@
...
 import sqlalchemy as sa
 from sqlalchemy.dialects import postgresql as psql
-from sqlalchemy.orm import backref, relationship, mapper
+from sqlalchemy.orm import relationship
 from sqlalchemy.dialects.postgresql import JSONB
 from sqlalchemy_utils import ArrowType
 
-from baselayer.app.models import (init_db, join_model, Base, DBSession, ACL,
-                                  Role, User, Token)
+from baselayer.app.models import (
```
Comment (acrellin) [inline skyportal/models.py]: `black`

File: skyportal/models.py [hunk 12]
```diff
@@ -123,79 +139,220 @@ class Source(Base):
...
-                            primaryjoin='group_users.c.user_id == users.c.id')
+User.sources = relationship(
+    "Source",
+    backref="users",
+    secondary="join(Group, group_sources).join(group_users)",
+    primaryjoin="group_users.c.user_id == users.c.id",
+)
+
+
+class Candidate(Base):
```
Comment (acrellin) [inline skyportal/models.py]: New candidates table

File: skyportal/models.py [hunk 13]
```diff
@@ -323,17 +510,33 @@ def from_ascii(cls, filename, source_id, instrument_id, observed_at):
 
 class Thumbnail(Base):
     # TODO delete file after deleting row
-    type = sa.Column(sa.Enum('new', 'ref', 'sub', 'sdss', 'dr8', "new_gz",
-                             'ref_gz', 'sub_gz',
-                             name='thumbnail_types', validate_strings=True))
+    type = sa.Column(
```
Comment (acrellin) [inline skyportal/models.py]: `black` (as are most other changes to this file)

File: skyportal/tests/api/test_candidates.py [hunk 14]
```diff
@@ -0,0 +1,199 @@
+import numpy.testing as npt
```
Comment (acrellin) [inline skyportal/tests/api/test_candidates.py]: API tests

File: skyportal/tests/api/test_thumbnail.py [hunk 15]
```diff
@@ -3,360 +3,444 @@
 import datetime
 import base64
 from skyportal.tests import api
-from skyportal.models import Thumbnail, DBSession, Photometry, Source
+from skyportal.models import Thumbnail, DBSession, Source
```
Comment (acrellin) [inline skyportal/tests/api/test_thumbnail.py]: Mostly `black`, with some bug fixes

File: skyportal/tests/frontend/test_candidate_scanning_page.py [hunk 16]
```diff
@@ -0,0 +1,272 @@
+import uuid
```
Comment (acrellin) [inline skyportal/tests/frontend/test_candidate_scanning_page.py]: Front-end tests

File: skyportal/tests/frontend/test_sources.py [hunk 17]
```diff
@@ -71,8 +65,8 @@ def test_download_comment_attachment(driver, user, public_source):
     assert(os.path.exists(fpath))
     try:
         with open(fpath) as f:
-            l = f.read()
-        assert l.split('\n')[0] == 'wavelength,flux,instrument_id'
+            lines = f.read()
```
Comment (acrellin) [inline skyportal/tests/frontend/test_sources.py]: Syntax

File: static/js/API.js [hunk 18]
```diff
@@ -77,8 +77,12 @@ function PUT(endpoint, actionType, payload) {
   return API(endpoint, actionType, 'PUT', payload);
 }
 
+function PATCH(endpoint, actionType, payload) {
```
Comment (acrellin) [inline static/js/API.js]: Add PATCH method

File: static/js/components/CandidateCommentList.jsx [hunk 19]
```diff
@@ -0,0 +1,43 @@
+import React from 'react';
```
Comment (acrellin) [inline static/js/components/CandidateCommentList.jsx]: Candidate comment list - different layout / functionality than the source comment list, so it gets its own component

File: static/js/components/FilterCandidateList.jsx [hunk 20]
```diff
@@ -0,0 +1,200 @@
+import React, { useState, useEffect } from 'react';
```
Comment (acrellin) [inline static/js/components/FilterCandidateList.jsx]: Filtering candidate list

File: static/js/components/SaveCandidateGroupSelectDialog.jsx [hunk 21]
```diff
@@ -0,0 +1,97 @@
+import React, { useState } from 'react';
```
Comment (acrellin) [inline static/js/components/SaveCandidateGroupSelectDialog.jsx]: Save candidate group select dialog

File: static/js/ducks/candidate.js [hunk 22]
```diff
@@ -0,0 +1,39 @@
+import * as API from '../API';
```
Comment (acrellin) [inline static/js/ducks/candidate.js]: candidate action creators/reducers/etc

File: static/js/ducks/candidates.js [hunk 23]
```diff
@@ -0,0 +1,57 @@
+import messageHandler from 'baselayer/MessageHandler';
```
Comment (acrellin) [inline static/js/ducks/candidates.js]: candidates action creators/reducers/etc

File: tools/load_demo_data.py [hunk 24]
```diff
@@ -3,15 +3,14 @@
...
-import shutil
 import pandas as pd
 import signal
 import requests
 
 from baselayer.app.env import load_env
 from baselayer.app.model_util import status, create_tables, drop_tables
 from social_tornado.models import TornadoStorage
-from skyportal.models import init_db, Base, DBSession, Source, User
+from skyportal.models import init_db, Base, Candidate, DBSession, Source, User
```
Comment (acrellin) [inline tools/load_demo_data.py]: add both saved and unsaved candidates during load_demo_data with associated photometry, spectra, thumbnails, etc...

File: static/js/components/CandidateList.jsx [hunk 25]
```diff
@@ -0,0 +1,137 @@
+import React, { useEffect } from 'react';
```
Comment (acrellin) [inline static/js/components/CandidateList.jsx]: This is the new candidate list component

**Comment by acrellin** [comment]
@stefanv did my review comments help? If you could use more explanation, I'd be happy to add to it.

**Comment by stefanv** [comment]
@acrellin Taking another look, I see that a lot of the changes were due to stylistic cleanups—let's try and do those in separate PRs in the future, so they don't grow the PRs unnecessarily.

I think in this case we can separate it out into the following changesets:

1. Database structure changes and `/api/candidates`-related changes
2. Cron job to keep the database clean
3. Frontend components and associated tests

Does that seem reasonably doable?  I can give it a shot, if it feels like I'm asking for too much.

**Comment by acrellin** [comment]
This is being broken up into several smaller PRs:
https://github.com/skyportal/skyportal/pull/382
https://github.com/skyportal/skyportal/pull/383
https://github.com/skyportal/skyportal/pull/384
https://github.com/skyportal/skyportal/pull/385
https://github.com/skyportal/skyportal/pull/386

**Pull request closed without merging**